posts/2023-06-29-linux-arm64-github-actions/index.qmd

```diff
@@ -0,0 +1,316 @@
+and the required LaTeX packages takes approximately 10 minutes on our emulated
+system, so there is a noticeable time benefit to using an image with it
+already installed.  The modified image can be found at
+[ghcr.io/grimbough/bioc-with-tinytex:devel-arm64](https://ghcr.io/grimbough/bioc-with-tinytex:devel-arm64)
+
+*Note: You could probably achieve a similar result by using the standard
+Bioconductor container and running `R CMD check` with the arguments `--no-manual`
+and `--no-build-vignettes`, however I would rather run the complete testing
+process in case there is problematic code in either the manual page examples or
+vignette.*
+
+## Installing packages
+
+Installing packages that require compilation is also incredibly slow on our emulated
+system, so it's immediately desirable to cache the library of installed packages
+that are needed for testing.  The [actions/cache](https://github.com/actions/cache) GitHub action does a good
+job of this, however it will only create a cache after a successful job run.
+Given we're creating this workflow to test a potentially problematic package, it
+can be frustrating to repeatedly wait several hours for all the necessary
+packages to install, because you haven't managed to fixed the issue.
+
+Given this, we can split our workflow into two jobs; the first installs the
+packages while the second runs the actual package tests.  With this structure a
+failure in the second job doesn't prevent the cache mechanism from working and
+makes repeated runs much faster.
+
+Now lets take a look at the first few steps on our `install-dependencies` job and
+explain what's happening.  Most of these steps are pretty standard for regular
+users of GitHub Actions.
+
+```yaml   
+  install-depdendencies:
+    name: Install package dependencies
+    runs-on: ubuntu-22.04
+    
+    steps:
+      - name: checkout
+        uses: actions/checkout@v3
+        
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
+        with:
+          platforms: arm64
+          
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v2
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+```
+
+First off we're checking out the git repository the workflow is found in.  That's
+probably just the package you're testing.  Then we're using the 
+[docker/setup-qemu-action](https://github.com/docker/setup-qemu-action) to
+install the QEMU emulator discussed in the [previous post](https://bioconductor.github.io/biocblog/posts/2023-06-09-debug-linux-arm64-on-docker/).  
+
+Next we login to the GitHub container registry in order to later pull the image
+were going to run the tests under.  For pulling container images you we can use the automatic [`GITHUB_TOKEN` secret](https://docs.github.com/en/actions/security-guides/automatic-token-authentication),
+so you don't need to configure anything here.
```

Comment:
Is this really necessary ?
https://github.com/grimbough/bioc-testing-with-arm64/pkgs/container/bioc-with-tinytex is a public package, so I think there is no need to login.

Reply:
I think you might be right, maybe that's a legacy from when I was testing with a private image.  I realised I don't do the log in step in the second job, so maybe it can all be deleted for simplicity.